### DISCUSSION TITLE
Store metrics using explicit schema

### DISCUSSION THREAD

**Discussion by ants** [cybertec-postgresql/pgwatch#447] [Ideas]
In pgw2 data and tags are stored as jsonb. This duplicates schema information in every row, giving bad storage efficiency for uncompressed data. It also defeats timeseries compression mechanisms for compressed data. Additionally, any access to data needs to decompress the whole document instead of accessing just the necessary column, which is very important for columnar storage.

Storing series metadata (e.g. full query text) out-of-line in a separate table might also be a good idea.

TBD: measurements on real world data

**Comment by ants** [comment]
Measurement results on 45M rows of real world stat_statements data (6db * 3mo) + a query to get month worth of data for a single queryid, 4 data cols.

| Schema     | Size uncompressed | Size compressed | Compression ratio | Query time | Buffers accessed |
| ---------- | ----------------- | --------------- | ----------------- | -----------| ---------------- |
| jsonb      | 70 GB             | 8050 MB         | 8.9x              | 1600ms     | 143596           |
| relational | 11 GB             | 763 MB          | 14.5x             | 4ms        | 522              |

**Comment by kmoppel** [comment]
My .02$ - in practice I think that extra data proliferation only becomes visible only with this `stat_statements` metric - and using TimescaleDB reduced old data heavily still (aroun ~10x) - so indeed, would be nice, but not sure if its worth the extra complexity as users on very high instance counts would probably prefer Prometheus anyways.
